### PR TITLE
[Fix] Fix GPTBigcoder for distributed execution

### LINKED ISSUE #488: AssertionError for WizardCoder-15B-V1.0 Again
2023.07.18
I pulled the latest code and ran:

`pip3 install . -e`
`python3 -c "from vllm import LLM;llm = LLM('/root//models/wizardcoder', tensor_parallel_size = 2)"`

and I got:

`AssertionError: transformer.h.0.attn.c_attn.weight shape mismatch between model and checkpoint: torch.Size([3200, 6144]) != torch.Size([3328, 6144])`

It seems that:
```
param.shape[0] = (6144 + 128 * 2) /  2 = 3200
loaded_weight.shape[0] = 6144 / 2 + 128 * 2 = 3328
```
under multi-query setting, which causes mismatch.

can you reproduce this? (or am I missing something)

### PR DESCRIPTION
**Pull request by zhuohan123**
Fix #488 and #487

### CODE REVIEW

**Comment by zhuohan123** [comment]
@WoosukKwon This PR is ready to review!

File: vllm/model_executor/models/gpt_bigcode.py [hunk 1]
```diff
@@ -291,9 +299,39 @@ def load_weights(self,
...
-                loaded_weight = torch.cat([wq, wk, wv], dim=0)
+            if name == "transformer.wte.weight":
+                # Consider padding in the vocab size.
+                padded_vocab_size = param.shape[
+                    0] * tensor_model_parallel_world_size
+                num_extra_rows = padded_vocab_size - self.config.vocab_size
+                extra_rows = torch.empty(num_extra_rows,
+                                         loaded_weight.shape[1])
+                extra_rows = extra_rows.to(loaded_weight)
+                loaded_weight = torch.cat([loaded_weight, extra_rows], dim=0)
```
Comment (WoosukKwon) [inline vllm/model_executor/models/gpt_bigcode.py:334]: Maybe not in this PR, but I think we can refactor this out, as it is used in multiple models.

**Review by WoosukKwon — APPROVED** [review]
LGTM! Thanks!

File: vllm/model_executor/models/gpt_bigcode.py (hunk 1 above)
Comment (zhuohan123) [reply vllm/model_executor/models/gpt_bigcode.py:334]: Sounds good! I think that makes a lot of sense. Also, the attention weights loading part can be refactored.

**Pull request merged**